cmd/ko/main.go

```diff
@@ -0,0 +1,266 @@
+		// from our commands.
+		FParseErrWhitelist: cobra.FParseErrWhitelist{
+			UnknownFlags: true,
+		},
+	})
+	fo := &FilenameOptions{}
+	apply := &cobra.Command{
+		Use: "apply -f FILENAME",
+		// TODO(mattmoor): Expose our own apply surface.
+		Short: `See "kubectl help apply" for detailed usage.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			// TODO(mattmoor): Use io.Pipe to avoid buffering the whole thing.
+			buf := bytes.NewBuffer(nil)
+			resolveFilesTo(fo, buf)
+
+			// Issue a "kubectl apply" command reading from stdin,
+			// to which we will pipe the resolved files.
+			kubectlCmd := exec.Command("kubectl", "apply", "-f", "-")
+
+			// Pass through our environment
+			kubectlCmd.Env = os.Environ()
+			// Pass through our std{out,err} and make our resolved buffer stdin.
+			kubectlCmd.Stderr = os.Stderr
+			kubectlCmd.Stdout = os.Stdout
+			kubectlCmd.Stdin = buf
+
+			// Run it.
+			if err := kubectlCmd.Run(); err != nil {
+				log.Fatalln(err)
+			}
+		},
+	}
+	addFileArg(apply, fo)
+	topLevel.AddCommand(apply)
+	resolve := &cobra.Command{
+		// TODO(mattmoor): Pick a better name.
+		Use:   "resolve -f FILENAME",
+		Short: `Print the input files with image references resolved to built/pushed image digests.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			resolveFilesTo(fo, os.Stdout)
+		},
+	}
+	addFileArg(resolve, fo)
+	topLevel.AddCommand(resolve)
+}
+
+func gobuildOptions() build.Options {
+	// A better story for O
+	base, err := remote.Image(baseImage, authn.Anonymous, http.DefaultTransport)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	return build.Options{
+		Base: base,
+	}
+}
+
+// From pkg/kubectl
+type FilenameOptions struct {
+	Filenames []string
+	Recursive bool
+}
+
+func addFileArg(cmd *cobra.Command, fo *FilenameOptions) {
+	// From pkg/kubectl
+	cmd.Flags().StringSliceVarP(&fo.Filenames, "filename", "f", fo.Filenames,
+		"Filename, directory, or URL to files to use to create the resource")
+	cmd.Flags().BoolVarP(&fo.Recursive, "recursive", "R", fo.Recursive,
+		"Process the directory used in -f, --filename recursively. Useful when you want to manage related manifests organized within the same directory.")
+}
+
+// Based heavily on pkg/kubectl
+func enumerateFiles(fo *FilenameOptions) ([]string, error) {
+	var files []string
+	for _, paths := range fo.Filenames {
+		err := filepath.Walk(paths, func(path string, fi os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if fi.IsDir() {
+				if path != paths && !fo.Recursive {
+					return filepath.SkipDir
+				}
+				return nil
+			}
+			// Don't check extension if the filepath was passed explicitly
+			if path != paths {
+				switch filepath.Ext(path) {
+				case ".json", ".yaml":
+					// Process these.
+				default:
+					return nil
+				}
+			}
+
+			files = append(files, path)
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return files, nil
+}
+
+func resolveFilesTo(fo *FilenameOptions, out io.Writer) {
```

Comment:
It feels like you stopped naming this function halfway through

Reply:
Done